modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/mapping/MappingServiceImpl.java

```diff
@@ -96,12 +108,54 @@ public CompletableFuture<List<MappedFragment>> map(MultiStepPlan multiStepPlan)
         return initialTopologyFuture.thenComposeAsync(ignore -> map0(multiStepPlan), taskExecutor);
     }
 
+    /** Called when the primary replica has expired. */
+    public CompletableFuture<Boolean> onPrimaryReplicaExpired(PrimaryReplicaEventParameters parameters) {
+        assert parameters != null;
+        assert parameters.groupId() instanceof TablePartitionId;
+
+        int tabId = ((TablePartitionId) parameters.groupId()).tableId();
+
+        return CompletableFuture.supplyAsync(() -> {
+            mappingsCache.values().removeIf(value -> value.tableIds.contains(tabId));
+
+            return false;
+        }, ForkJoinPool.commonPool());
```

Comment:
It looks like the full scan should be done asynchronously, at least to avoid blocking other listeners. But I'm not sure which pool is best to use for this.

I prefer not to use the query execution pool for this. Because firstly, there may be quite long tasks in the queue, and secondly, we can slow it down with idle cache scanning tasks.

Reply:
Result of benchmark for 10 times scan with random `removeIf(...)


| Benchmark   |    (cacheSize) | (tablesInPlan) |  (totalTables) |     Score   |    Error |  Units |
| ------------|----------------|----------------|----------------|-------------|----------|------- |
| 1 thread               |1024              |10          |10000  |    317.112 ±    |29.731  |us/op
| 8 threads              |1024              |10          |10000  |    286.909 ±     |5.741  |us/op
| 8 threads + sync       |1024              |10          |10000  |    294.775 ±     |4.226  |us/op
| ------------|----------------|----------------|----------------|-------------------------------- |
|1 thread               |1024             |100          |10000   |   340.265 ±    |15.269  |us/op
|8 threads              |1024             |100          |10000   |   287.221 ±    | 3.058  |us/op
|8 threads + sync       |1024             |100          |10000   |   349.131 ±    |49.229  |us/op
| ------------|----------------|----------------|----------------|-------------------------------- |
| 1 thread              |10000              |10          |10000  |   3136.671 ±   |596.822  |us/op
| 8 threads             |10000              |10          |10000  |   2822.520 ±   |280.104  |us/op
| 8 threads + sync      |10000              |10          |10000  |   3252.009 ±   |205.003  |us/op
| ------------|----------------|----------------|----------------|-------------------------------- |

| 1 thread              |10000             |100          |10000  |   3586.382 ±   |159.551  |us/op
| 8 threads             |10000             |100          |10000  |   3453.345 ±   |104.515  |us/op
| 8 threads + sync      |10000             |100          |10000  |   3641.002 ±   |192.040  |us/op
| ------------|----------------|----------------|----------------|-------------------------------- |

| 1 thread             |100000              |10          |10000  |  40390.912 ±  |5187.521  |us/op
| 8 threads            |100000              |10          |10000  |  38434.787 ±  |8073.539  |us/op
| 8 threads + sync     |100000              |10          |10000  |  41262.388 ±  |8081.194  |us/op
| ------------|----------------|----------------|----------------|-------------------------------- |
| 1 thread             |100000             |100          |10000  |  63233.467 ± |12400.052  |us/op
| 8 threads            |100000             |100          |10000  |  64562.198 ±  |7600.780  |us/op
| 8 threads + sync     |100000             |100          |10000  |  63495.429 ±  |6680.330  |us/op
| ---------------------|----------------|----------------|-------|---------------|----------|---------|

Reply:
Benchmark scenario:
* create cache + fill cache
* measure the time of 10 calls `mappingsCache.values().removeIf(value -> value.tableIds.contains(RANDOM_TABLE_ID));`

Measuring time for single thread and for multiple threads.
`8 threads + sync` means that the `mappingsCache.values().removeIf(value -> value.tableIds.contains(RANDOM_TABLE_ID));` is enclosed in a synchronized block (to reduce possible contention when scanning).

`tablesInPlan` - size of set `tableIds`.
`totalTables` - total number of table IDs (the lower this number, the more cache elements will be deleted in one scan)

```
Benchmark     (cacheSize)  (tablesInPlan)  (totalTables)  Mode  Cnt       Score       Error  Units
1 thread             1024              10          10000  avgt    8     298.156 ±    16.015  us/op
8 threads            1024              10          10000  avgt    8     401.057 ±    14.550  us/op
8 threads + sync     1024              10          10000  avgt    8    1577.090 ±    77.315  us/op

1 thread             1024             100          10000  avgt    8     318.541 ±    12.416  us/op
8 threads            1024             100          10000  avgt    8     418.963 ±    30.124  us/op
8 threads + sync     1024             100          10000  avgt    8     878.342 ±    92.320  us/op

1 thread            10000              10          10000  avgt    8    2974.037 ±   115.337  us/op
8 threads           10000              10          10000  avgt    8    4595.802 ±   282.642  us/op
8 threads + sync    10000              10          10000  avgt    8   15558.993 ±  1107.365  us/op

1 thread            10000             100          10000  avgt    8    3481.707 ±    37.420  us/op
8 threads           10000             100          10000  avgt    8    4652.641 ±   124.261  us/op
8 threads + sync    10000             100          10000  avgt    8    8283.897 ±   588.364  us/op

1 thread           100000              10          10000  avgt    8   37814.228 ±  3901.386  us/op
8 threads          100000              10          10000  avgt    8   56042.789 ±  2641.133  us/op
8 threads + sync   100000              10          10000  avgt    8  170200.745 ± 50732.738  us/op

1 thread           100000             100          10000  avgt    8   62899.692 ± 11203.015  us/op
8 threads          100000             100          10000  avgt    8   66636.729 ±  6559.783  us/op
8 threads + sync   100000             100          10000  avgt    8  159911.761 ± 54040.121  us/op
```


The results obtained on my machine show that there are no contention between threads during this operation (or it is not significant).